File: ios/Model/Toast.swift
//
//  ToastModel.swift
//  NitroToast
//
//  Created by kiet.huynh on 6/3/25.
//

import Foundation
import SwiftUI

enum ToastIcon {
  case system(name: String, color: Color)
  case image(uri: String)
  case progress(color: Color)
  case none
}

final class Toast: Identifiable, ObservableObject {
  var id: String
  @Published var message: String
  @Published var config: NitroToastConfig

  /// View Properties
    @Published var isUpdating: Bool = false
  @Published var isPaused: Bool = false
  @Published var isDeleting: Bool = false

  init(toastId: String, message: String, config: NitroToastConfig) {
    self.id = toastId
    self.message = message
    self.config = config
  }
}

extension NitroToastConfig: Equatable {
  public static func == (lhs: NitroToastConfig, rhs: NitroToastConfig) -> Bool {
    return lhs.presentation == rhs.presentation && lhs.type == rhs.type
  }
}

extension Toast {
  var backgroundColor: Color {
    if let hex = config.backgroundColor {
      return Color(hex)
    }

    switch config.type {
    case .success:
      return Color.toastSuccess
    case .error:
      return Color.toastError
    case .info:
      return Color.toastInfo
    case .warning:
      return Color.toastWarning
    case .loading:
      return Color.toastDefault
    case .default:
      return Color.toastDefault
    }
  }

  var overlayColor: Color {
    if config.useOverlay {
      return self.backgroundColor.opacity(0.08)
    }
    return self.backgroundColor
  }

  var icon: ToastIcon {
    if let uri = config.iconUri {
      return .image(uri: uri)
    }

    switch config.type {
    case .success:
      return .system(name: "checkmark.circle", color: iconColor)
    case .error:
      return .system(name: "exclamationmark.circle", color: iconColor)
    case .info:
      return .system(name: "info.circle", color: iconColor)
    case .warning:
      return .system(name: "exclamationmark.triangle", color: iconColor)
    case .loading:
      return .progress(color: iconColor)
    case .default:
      return .system(name: "bell", color: iconColor)
    }
  }
  var iconColor: Color {
    if config.useOverlay {
      return self.backgroundColor
    }
    return .white
  }

  var title: String {
    if let title = config.title {
      return title
    }
    switch config.type {
    case .success:
      return "Success"
    case .error:
      return "Error Occurred"
    case .info:
      return "Information"
    case .warning:
      return "Warning"
    case .loading:
      return "Loading..."
    case .default:
      return "Notice"
    }
  }

  var titleColor: Color {
    if let hex = config.titleColor {
      return Color(hex)
    }
    return .black
  }

  var messageColor: Color {
    if let hex = config.messageColor {
      return Color(hex)
    }
    return .secondary
  }
}

File: android/src/main/java/com/margelo/nitro/nitrotoast/Toast.kt
package com.margelo.nitro.nitrotoast

import androidx.compose.ui.graphics.Color
import androidx.core.graphics.toColorInt

fun String.toColorOrNull(): Color? =
    try {
        Color(this.toColorInt())
    } catch (e: IllegalArgumentException) {
        null // or log error
    }

data class Toast(
    val id: String,
    val message: String,
    val config: NitroToastConfig,
    var offsetX: Float = 0f,
    var isVisible: Boolean = false,
    var isPaused: Boolean = false,
    var isUpdating: Boolean = false,
) {
    val backgroundColor: Color
        get() =
            config.backgroundColor?.toColorOrNull() ?: when (config.type) {
                ToastType.SUCCESS -> Color(0xFF22C55E)
                ToastType.ERROR -> Color(0xFFEF4444)
                ToastType.INFO -> Color(0xFF3B82F6)
                ToastType.WARNING -> Color(0xFFF59E0B)
                ToastType.LOADING -> Color(0xFF6B7280)
                ToastType.DEFAULT -> Color(0xFF6B7280)
            }

    val overlayColor: Color
        get() = if (config.useOverlay) backgroundColor.copy(alpha = 0.08f) else backgroundColor

    val icon: ToastIcon
        get() {
            config.iconUri?.let { return ToastIcon.Image(it) }
            return when (config.type) {
                ToastType.SUCCESS -> ToastIcon.System("checkmark", iconColor)
                ToastType.ERROR -> ToastIcon.System("error", iconColor)
                ToastType.WARNING -> ToastIcon.System("warning", iconColor)
                ToastType.INFO -> ToastIcon.System("info", iconColor)
                ToastType.LOADING -> ToastIcon.Progress(iconColor)
                ToastType.DEFAULT -> ToastIcon.System("bell", iconColor)
            }
        }
    private val iconColor: Color
        get() = if (config.useOverlay) backgroundColor else Color.White

    val title: String
        get() =
            config.title ?: when (config.type) {
                ToastType.SUCCESS -> "Success"
                ToastType.ERROR -> "Error Occurred"
                ToastType.INFO -> "Information"
                ToastType.WARNING -> "Warning"
                ToastType.LOADING -> "Loading..."
                ToastType.DEFAULT -> "Notice"
            }

    val titleColor: Color
        get() = config.titleColor?.toColorOrNull() ?: Color.Black

    val messageColor: Color
        get() = config.messageColor?.toColorOrNull() ?: Color.DarkGray
}
